### PR TITLE
ARROW-2117: [C++] Update codebase / CI toolchain for clang 5.0

### PR DESCRIPTION
**Pull request by wesm**
LLVM 5.0 found a number of code tidying issues that I fixed here. It also wants us to use `override` with our virtual destructor implementations, which I wasn't aware of whether is a best practice or not. 

cc @xhochy @cpcloud for review

### CODE REVIEW

File: cpp/src/arrow/memory_pool.h [hunk 1]
```diff
@@ -71,7 +71,7 @@ class ARROW_EXPORT MemoryPool {
 class ARROW_EXPORT LoggingMemoryPool : public MemoryPool {
  public:
   explicit LoggingMemoryPool(MemoryPool* pool);
-  virtual ~LoggingMemoryPool() = default;
+  ~LoggingMemoryPool() override = default;
```
Comment (cpcloud) [inline cpp/src/arrow/memory_pool.h]: Overridden default destructors can be removed. Same goes for the empty destructors.

Comment (cpcloud) [inline cpp/src/arrow/memory_pool.h]: Unless they are base classes of course :)

File: cpp/src/arrow/array-test.cc [hunk 2]
```diff
@@ -1287,7 +1288,7 @@ TEST_F(TestFWBinaryArray, Builder) {
 
   std::shared_ptr<Array> result;
 
-  auto CheckResult = [&length, &is_valid, &raw_data, byte_width](const Array& result) {
+  auto CheckResult = [&length, &is_valid, &raw_data](const Array& result) {
```
Comment (wesm) [inline cpp/src/arrow/array-test.cc]: TIL you can reference local `constexpr` values inside lambdas without declaring in the capture clause

File: cpp/src/arrow/memory_pool.h (hunk 1 above)
Comment (xhochy) [reply cpp/src/arrow/memory_pool.h]: If they are base classes, they should not need override.

Comment (cpcloud) [reply cpp/src/arrow/memory_pool.h]: I meant that you can't remove virtual destructors from base classes.

**Comment by cpcloud** [comment]
I'm looking into the build failures now, so we can get this merged. Looks like there some incorrect package spec somewhere in the CI scripts maybe.

**Comment by cpcloud** [comment]
Looks like we may just need to call `apt-get update`. Taking a closer look.

**Comment by cpcloud** [comment]
What a completely insane Heisenbug. Apparently enough clicking solves the problem.

**Review by cpcloud — APPROVED** [review]
LGTM. @wesm Is this ready to be merged?

**Comment by wesm** [comment]
SGTM. we should notify folks on the mailing list to upgrade their Clang toolchains to 5.0

**Comment by cpcloud** [comment]
Cool, merging.

**Pull request closed without merging**